src/translate/pddl_parser/parsing_functions.py

```diff
@@ -1,30 +1,103 @@
+import contextlib
 import sys
 
 import graph
 import pddl
+from .parse_error import ParseError
+
+
+class Context:
+    def __init__(self):
+        self._traceback = []
+
+    def check_string(self, item, value=None, finalmsg=None, item_name="Element"):
+        finalmsg = '' if finalmsg is None else finalmsg
+        if not isinstance(item, str):
+            self.error(f"{item_name} is expected to be a word. "
+                       f"{finalmsg}\nGiven: {item}")
+        if value and item != value:
+            self.error(f"{item_name} is expected to have the value '{value}'.\n "
+                       f"Given: {item}")
+
+    def check_list(self, item, min_size=None, max_size=None,
+                   block_name="Block", item_name="element", finalmsg=None):
+        finalmsg = '' if finalmsg is None else finalmsg
+        if not isinstance(item, list):
+            self.error(f"{block_name} is expected to be a block. {finalmsg}")
+        if min_size and min_size == max_size and len(item) != min_size:
+            self.error(f"{block_name} should contain exactly {min_size} {item_name}(s). "
+                       f"{finalmsg}")
+        elif min_size and len(item) < min_size:
+            self.error(f"{block_name} should contain at least {min_size} {item_name}(s). "
+                       f"{finalmsg}")
+        elif max_size and len(item) > max_size:
+            self.error(f"{block_name} should contain at most {max_size} {item_name}(s). "
+                       f"{finalmsg}")
+
+    def check_expected_block(self, item, expected, min_size=1, max_size=None):
+        inner = expected[0] if len(expected) == 1 else ", ".join(expected)
```

Comment:
1. removed the unnecessary join.
2. changed syntax to this kind of output `Expecting '({:pddl, :predicates} ...)'.`  I use `(`, `)` with their literal meaning (PDDL is full of parentheses) and `|` it resembels regex syntax where `(`, `)` has a different mearning. Thus my syntax is
`{a,b,c}` for options and
`[asdf]` for a block, e.g. [asdf]?` for an optional block

Reply:
I would avoid introducing a formal syntax for the error messages. This will end up being confusing in some cases.
How about `Expecting a list where the first entry is one of :pddl, :predicates` or joining with ` or `: `Expecting a list where the first entry is :pddl or :predicates`